onnxruntime/core/providers/rocm/cu_inc/common.cuh

```diff
@@ -0,0 +1,243 @@
+__device__ __inline__ half _Tanh(half a) { return half(tanhf((float)a)); }
+
+template <>
+__device__ __inline__ half2 _Tanh(half2 a) {
+  float2 tmp = (__half22float2(a));
+  tmp.x = tanhf(tmp.x);
+  tmp.y = tanhf(tmp.y);
+  return __float22half2_rn(tmp);
+}
+
+// Capture permutations of int32/64/float/double
+template <typename T, typename T1>
+__device__ __inline__ T _Pow(T a, T1 b) {
+  return static_cast<T>(pow(static_cast<double>(a), static_cast<double>(b)));
+}
+
+template <>
+__device__ __inline__ float _Pow(float a, float b) { return powf(a, b); }
+
+template <>
+__device__ __inline__ double _Pow(double a, double b) { return pow(a, b); }
+
+template <>
+__device__ __inline__ half _Pow(half a, half b) { return half(powf((float)a, (float)b)); }
+
+template <typename T>
+__device__ __inline__ T _Min(T a, T b) { return a < b ? a : b; }
+
+template <typename T>
+__device__ __inline__ T _Max(T a, T b) { return a > b ? a : b; }
+
+template <typename T>
+__device__ __inline__ T _Abs(T a) { return a > (T)0 ? a : -a; }
+
+template <typename T>
+__device__ __inline__ T _Normcdf(T a);
+
+template <>
+__device__ __inline__ float _Normcdf(float a) { return normcdff(a); }
+
+template <>
+__device__ __inline__ double _Normcdf(double a) { return normcdf(a); }
+
+template <>
+__device__ __inline__ half _Normcdf(half a) { return half(normcdff((float)a)); }
+
+template <typename T>
+__device__ __inline__ T _Gelu(T a) {
+  return a * _Normcdf(a);
+}
+
+
+// We would like to use 64-bit integer to support large matrices. However, HIP seems to support only 32-bit integer
+// For now, use int32_t to ensure that both Linux and Windows see this as 32 bit integer type.
+#ifndef HIP_LONG
+#define HIP_LONG int32_t
+#endif
+
+template <class INT, class INT2>
+static INT CeilDiv(INT a, INT2 b)  // ceil(a/b)
+{
+  return (INT)(((size_t)a + (size_t)b - 1) / (size_t)b);  // these size_t casts are necessary since b may be INT_MAX (for maxGridSize[])
+}
+
+struct GridDim {
+  enum : HIP_LONG {
+    maxThreadsPerBlock = 256,  // max threads per block
+    maxElementsPerThread = 4,  // max element processed per thread
+  };
+};
+
+
+#define CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(id, N)          \
+  HIP_LONG id = blockDim.x * blockIdx.x + threadIdx.x;     \
+  if (id >= N)                                              \
+    return;
+
+// HIP_KERNEL_ASSERT is a macro that wraps an assert() call inside hip kernels.
+// This is not supported by Apple platforms so we special case it.
+// See http://docs.nvidia.com/rocm/hip-c-programming-guide/#assertion
+#if defined(__APPLE__) || defined(__HIP_PLATFORM_HCC__)
+#define HIP_KERNEL_ASSERT(...)
+#else // __APPLE__
+#define HIP_KERNEL_ASSERT(...) assert(__VA_ARGS__)
+#endif // __APPLE__
+
+// WARP related definitions and functions
+constexpr int GPU_WARP_SIZE = 64;
+
+template <typename T>
+__device__ __forceinline__ T WARP_SHFL(T value, int srcLane, int width = GPU_WARP_SIZE, unsigned int mask = 0xffffffff)
+{
+#if HIP_VERSION >= 9000
```

Comment:
The version number 9000 matches for CUDA. Is that correct?

Reply:
it is not correct. It should be removed.